src/persist-client/src/impl/shard.rs

```diff
@@ -206,6 +206,47 @@ where
         Ok(new_upper)
     }
 
+    pub async fn compare_and_append<'i, I: IntoIterator<Item = ((&'i K, &'i V), &'i T, &'i D)>>(
+        &self,
+        writer_id: &WriterId,
+        updates: I,
+        expected_upper: Antichain<T>,
+        new_upper: Antichain<T>,
+    ) -> Result<Result<Antichain<T>, Antichain<T>>, InvalidUsage> {
+        let mut state = self.state.lock().await;
+
+        match state.writers.get(writer_id) {
+            Some(_) => (), // The writer is known
+            // TODO: This is more likely an expired lease.
+            None => return Err(InvalidUsage(anyhow!("unknown writer: {:?}", writer_id))),
+        };
+
+        let lower = Self::decode_antichain(&state.upper);
+
+        if PartialOrder::less_than(&new_upper, &lower) {
```

Comment:
proposal: the way I think of the validations for this are 1) expected_upper vs new_upper (deterministic and a failure is an invalid usage) 2) expected_upper vs shard_upper (a cas failure). what you have is (I think) correct but slightly less intuitive (at least for me). thoughts?

Reply:
I believe you are correct, this one check didn't do anything useful because the second check (for the expected upper) already covers this.

Reply:
sorry if i was unclear, but I think both your checks are necessary and equivalent to mine. I was just presenting a possible reframing